### PR TITLE
Prevent automatic ES service software upgrades (#3959)

### PR DESCRIPTION
**Pull request by achave11**
Connected issue: #3959

### Author

- [x] PR title references the connected issue
- [x] PR title matches<sup>1</sup> title of connected issue         <sub>or comment in PR explains why they're different</sub>
- [x] Title of at least one commit references connected issue
- [x] PR is connected to issue via Zenhub 
- [x] PR description links to connected issue

<sup>1</sup> when the issue title describes a problem, the PR title is `Fix: ` followed by the issue title   

### Author (reindex)

- [x] Added `r` tag to commit title                                 <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                                   <sub>or this PR does not require reindexing</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain                <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                        <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst          <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                                 <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                                   <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                          <sub>or this PR does not require announcement</sub>
- [x] Added checklist items for additional operator tasks           <sub>or this PR does not require additional tasks</sub>

## Author (hotfixes)

- [x] Added `h` tag to commit title and PR targets `prod`           <sub>or this PR does not include a temporary hotfix</sub>
- [x] Added `H` tag to commit title                                 <sub>or this PR does not include a permanent hotfix</sub>
- [x] Added `hotfix` label to PR                                    <sub>or this PR does not include a hotfix</sub>
- [x] Reverted the temporary hotfix connected to the issue          <sub>or there is no temporary hotfix for the connected issue on the `prod` branch</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                                <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                                 <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                                      <sub>or this PR does not touch requirements*.txt</sub>

### Author (before every review)

- [x] `make integration_test` passes in personal deployment         <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issue as `demo` or `no demo`
- [x] Commented on connected issue about demo expectations          <sub>or labelled connected issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Ticket is in *Review requested* column
- [x] Requested review from peer reviewer
- [x] Assigned PR to peer reviewer

### Peer reviewer (after approval)

- [x] `N reviews` label on PR reflects prior peer reviews with changes requested
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear                      <sub>or connected issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label             <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                                       <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                                  <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                             <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issue to Merged column
- [ ] Pushed merge commit to Github

### Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                         <sub>or PR description does not contain an announcement</sub>
- [ ] Shortened the PR chain                                        <sub>or this PR is not the base of another PR</sub>
- [ ] Pushed merge commit to Gitlab                                 <sub>or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Build passes on Gitlab
- [ ] Moved connected issue to `prod` or `Merged prod`              <sub>or this PR does not represent a promotion</sub>

### Operator (reindex) 

- [ ] Started reindex in target deployment                          <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures                     <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment                      <sub>or this PR does not require reindexing</sub>
- [ ] Filed backport PR                                             <sub>or this PR does not represent a hotfix</sub>

### Operator

- [ ] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

File: OPERATOR.rst [hunk 1]
```diff
@@ -55,6 +55,27 @@ Getting started as operator
...
+upgrade is received. The operator must prevent this upgrade from happening
+automatically and instead manually apply it through the AWS OpenSearch Service
+console when prod is idle.
+
+To check for upgrades:
+
+#. Go to https://us-east-1.console.aws.amazon.com/esv3/home?region=us-east-1#opensearch/notifications
+
+#. If notifications with subject `Service Software Update` are present, open a
+   new issue in https://github.com/DataBiosphere/azul/issues/new/choose and
```
Comment (hannes-ucsc) [inline OPERATOR.rst:72]: same here re link but everyone knows how to open an issue

File: OPERATOR.rst [hunk 2]
```diff
@@ -55,6 +55,27 @@ Getting started as operator
...
+
+Amazon OpenSearch Service releases service software updates periodically, which
+are automatically applied after two weeks from the time a notification about the
+upgrade is received. The operator must prevent this upgrade from happening
+automatically and instead manually apply it through the AWS OpenSearch Service
+console when prod is idle.
+
+To check for upgrades:
+
+#. Go to https://us-east-1.console.aws.amazon.com/esv3/home?region=us-east-1#opensearch/notifications
```
Comment (hannes-ucsc) [inline OPERATOR.rst:69]: This is not how links are formatted in rst.

Anyways: don't paste deep links to the AWS console. They break in no time. Use English to describe where to go.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
You gotta think this thru a bit more with respect to the different ES domains. You gotta take sandbox into account. I don't think we want a ticket for each deployment since I am pretty sure the update notification pretty much arrive at the same time. We should apply the update on `dev` first, then `sandbox` after making sure that no one is using it. Prod should be updated last.

File: OPERATOR.rst [hunk 3]
```diff
@@ -55,6 +55,57 @@ Getting started as operator
...
 -------------
 
+Check weekly for Amazon OSS (ES) Updates
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Check weekly for notifications about Amazon OpenSearch Service (OSS) software
+updates. Notifications with `High` severity will apply the update automatically
+after two weeks from the date the notification was received, read
+`Amazon notification severities`_ for more info. The operator should
+make the update manually as part of a task of an Azul issue.
```
Comment (achave11) [inline OPERATOR.rst:65]:
```suggestion
make the update manually as part of an operator task.
```

File: OPERATOR.rst [hunk 4]
```diff
@@ -55,6 +55,57 @@ Getting started as operator
...
+
+Check for Amazon OSS software updates:
+
+#. In ``Amazon OpenSearch Service Console`` select the ``Notifications`` pane
+   and identify notifications with subject ``'Service Software Update'``.
+
+#. Record the severity e.g. ``"Informational" time utc e.g
+   ``"3/22/2022, 11:06:14 PM"`` and domain name e.g. ``"azul-index-sandbox"`` of
+   these notifications. Collect this information for both ``prod`` and ``dev``
+   AWS accounts.
```
Comment (achave11) [inline OPERATOR.rst:75]:
```suggestion
#. Record the severity e.g. ``"Informational", time utc e.g
   ``"3/22/2022, 11:06:14 PM"``, and domain name e.g. ``"azul-index-sandbox"`` of
   these notifications. Collect this information for both ``prod`` and ``dev``
   AWS accounts.
```

File: OPERATOR.rst [hunk 5]
```diff
@@ -55,6 +55,57 @@ Getting started as operator
...
+
+#. In ``Amazon OpenSearch Service Console`` select the ``Notifications`` pane
+   and identify notifications with subject ``'Service Software Update'``.
+
+#. Record the severity e.g. ``"Informational", time utc e.g
+   ``"3/22/2022, 11:06:14 PM"``, and domain name e.g. ``"azul-index-sandbox"`` of
+   these notifications. Collect this information for both ``prod`` and ``dev``
+   AWS accounts.
+
+#. Open a new issue in Azul and title it ``Service Software Update {date}``.
```
Comment (achave11) [inline OPERATOR.rst:77]:
```suggestion
#. Open a new issue in Azul and title it ``Apply OpenSearch (ES) Service Software Update before {date}``.
```

File: OPERATOR.rst [hunk 6]
```diff
@@ -55,6 +55,58 @@ Getting started as operator
...
+   ``Apply Amazon OSS (ES) Software Update (before {date})``.
+   Title should include ``before {date}`` only if any recorded notification is
+   of ``High`` severity. {date} should be two weeks from the date for time utc.
+   e.g. ``Apply Amazon OSS (ES) Software Update before 4/05/2022``.
+
+#. Pin ticket with `High Priority` in ZenHub only if title contains a date,
+   i.e., a notification of ``High`` severity was recorded.
+
+#. The description of the newly created issue should include a checklist item
+   for each domain name recorded in step 2. Be considered of disruptions to
```
Comment (hannes-ucsc) [inline OPERATOR.rst:88]:
```suggestion
   for each domain name recorded in step 2. Be considerate of disruptions to
```

But discuss in PL.

File: OPERATOR.rst [hunk 7]
```diff
@@ -55,6 +55,58 @@ Getting started as operator
...
+   others workflow and incompatible upgrades when producing the checklist. The
+   checklist should reflect a sensible order for performing the upgrades. Always
```
Comment (hannes-ucsc) [inline OPERATOR.rst:90]: It's not clear to the reader what this means. Discuss in PL

File: OPERATOR.rst [hunk 8]
```diff
@@ -55,6 +55,58 @@ Getting started as operator
...
+   confirm with lead about adding unaccounted domain names to the checklist.
+   Checklist::
+
+    - [ ] Update `azul-index-dev`
+    - [ ] Confirm with Azul devs that their personal deployments are inactive
+    - [ ] Update `azul-index-sandbox`
+    - [ ] Update `azul-index-prod`
+
+#. To update the domain, click the desired domain in the OSS console. Under
+   ``General information`` the ``Service software version`` should have a
```
Comment (hannes-ucsc) [inline OPERATOR.rst:100]:
```suggestion
   ``General information``, the ``Service software version`` should have a
```

File: OPERATOR.rst [hunk 9]
```diff
@@ -55,6 +55,58 @@ Getting started as operator
...
+   ``Update available`` hyper link, click on it and complete the prompted
+   instructions.
+
+#. Once the upgrade process is completed, do a smoke test in the Data Browser to
```
Comment (hannes-ucsc) [inline OPERATOR.rst:104]: For `dev` and `prod`. Can't be done for `sandbox`

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: OPERATOR.rst [hunk 10]
```diff
@@ -55,6 +55,59 @@ Getting started as operator
 Operator jobs
 -------------
 
+Check weekly for Amazon OpenSearch Service updates
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Check weekly for notifications about Amazon OpenSearch Service (ElasticSearch
+[ES]) software updates. Notifications with `High` severity will be subject to an
```
Comment (achave11) [inline OPERATOR.rst:62]:
```suggestion
Check weekly for notifications about Amazon OpenSearch Service
software updates. Notifications with `High` severity will be subject to an
```

File: OPERATOR.rst [hunk 11]
```diff
@@ -55,6 +55,59 @@ Getting started as operator
...
+automatic update by AWS after two weeks from the date received. Read
+`Amazon notification severities`_ for more info. The operator should avoid any
```
Comment (achave11) [inline OPERATOR.rst:64]:
```suggestion
`Amazon notification severities`_ for more info. The operator should prevent any
```

File: OPERATOR.rst [hunk 12]
```diff
@@ -55,6 +55,59 @@ Getting started as operator
...
+automatic installation of updates. It would be disastrous if an update were to
+be applied in the middle of a reindex. Instead, the operator must apply the
+update manually as part of an operator ticket, as soon as possible, and well
+before any updates would be applied automatically.
+
+Check for Amazon OpenSearch software updates:
+
+#. In *Amazon OpenSearch Service Console* select the *Notifications* pane
+   and identify notifications with subject ``'Service Software Update'``.
```
Comment (achave11) [inline OPERATOR.rst:73]:
```suggestion
   and identify notifications with subject ``Service Software Update``.
```
here and elsewhere.

File: OPERATOR.rst [hunk 13]
```diff
@@ -55,6 +55,59 @@ Getting started as operator
...
+
+#. Record the severity e.g. ``"Informational"``, time utc e.g
+   ``"3/22/2022, 11:06:14 PM"``, and domain name e.g. ``"azul-index-sandbox"`` of
+   these notifications. Collect this information for both ``prod`` and ``dev``
+   AWS accounts.
```
Comment (achave11) [inline OPERATOR.rst:78]: Remove examples.

File: OPERATOR.rst [hunk 14]
```diff
@@ -55,6 +55,59 @@ Getting started as operator
...
+
+#. Open a new issue in Azul and title it
+   ``Apply Amazon OpenSearch (ES) Software Update (before {date})``.
+   Title should include ``(before {date})`` only if any recorded notification is
+   of ``High`` severity. {date} should be two weeks from the date that the
+   notification with high severity was received. e.g. ``Apply Amazon OpenSearch
+   (ES) Software Update before 4/05/2022``.
+
+#. Pin ticket with *High Priority* in ZenHub only if title contains a date,
+   i.e., a notification of ``High`` severity was recorded.
```
Comment (achave11) [inline OPERATOR.rst:88]:
```suggestion
#. Pin ticket with *High Priority* in ZenHub only if title contains a date.
```

File: OPERATOR.rst [hunk 15]
```diff
@@ -55,6 +55,59 @@ Getting started as operator
...
+
+#. The description of the newly created issue should include a checklist item
+   for each domain name recorded in step 2. The checklist should reflect items
+   for notifying developers of the incoming disruptions to the shared domains.
```
Comment (achave11) [inline OPERATOR.rst:92]:
```suggestion
   for notifying developers of any disruptions to their deployments.
```

File: OPERATOR.rst [hunk 16]
```diff
@@ -55,6 +55,59 @@ Getting started as operator
...
+   Checklist::
+
+    - [ ] Update `azul-index-dev`
+    - [ ] Confirm with Azul devs that their personal deployments are inactive
```
Comment (achave11) [inline OPERATOR.rst:96]:
```suggestion
    - [ ] Confirm with Azul devs that their personal deployments are idle
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Review by amarjandu — APPROVED** [review]

**Pull request merged**